### ISSUE TITLE
event_queue_remove_active:  not on queue 8

### ISSUE THREAD

**Issue by Zourva** [libevent/libevent#469]
Hi all, 
we've implemented an rpc server based on evhtp + libevent2 and we have many http clients connecting to our rpc server. 

We start from 10 clients to 2000, and the rpc server works okay. However, when we increase the number of clients to 2000+ ( or sometimes 3000+) it fails, printing out the log: `event_queue_remove_active: 0x??? not on queue 8`,  and exit with `-1`. 

We tried two versions(`libevent-release-2.0.22-stable` & `libevent-2.1.8-stable`), both show the error. We also tried the debug version and it strangely didn't show up. However, after we read the code we found the execution flow of the debug version is not the same as that of the relase one, though they used the same optimization level(`-O2`).  So we modified the logging functions and added some log prints at places like `event_add`/`event_del`/`queue_remove`/`insert` and compile it as release and run.  The error shows again and here's all the output: 

** backtrace:**
```c
#0  0x000000000602a5e5 in raise () from /lib64/libc.so.6
#1  0x000000000602bdc5 in abort () from /lib64/libc.so.6
#2  0x000000000602370e in __assert_fail_base () from /lib64/libc.so.6
#3  0x00000000060237d0 in __assert_fail () from /lib64/libc.so.6
#4  0x000000000763d1ca in fatal_cb (err=-559030611) at xxxx.cpp:81
#5  0x0000000007c5d291 in event_exit (errcode=-559030611) at log.c:102
#6  0x0000000007c5d35c in event_errx (eval=-559030611, fmt=<value optimized out>) at log.c:162
#7  0x0000000007c4fe7d in event_queue_remove_active (base=0x6c557e0, evcb=0xaae5410) at event.c:3200
#8  0x0000000007c54fc4 in event_process_active_single_queue (base=0x6c557e0, activeq=0x6c62020, max_to_process=2147483647, endtime=0x0) at event.c:1606
#9  0x0000000007c55aef in event_process_active (base=0x6c557e0, flags=<value optimized out>) at event.c:1738
#10 event_base_loop (base=0x6c557e0, flags=<value optimized out>) at event.c:1961
...
```

(NOTE: line numbers may not be the same as the original code as we've added some logging codes)

-------------------------------------------------------------------
**modified log functions and logs:**
```c
#ifdef EVENT_DEBUG_LOGGING_ENABLED
#define event_debug(x) do {			\
	if (event_debug_get_logging_mask_()) {	\
		event_debugx_ x;		\
	}					\
	} while (0)
#else
#define event_debug(x) ((void)0)
#endif
```

--- our own log functions
```c
#define event_user_debug(x) do {                     \
                event_debugx_ x;                \
        } while (0)

#define event_user_test_log_evcb(evcb, tv) do{                                                         \
    struct timeval cur_tv;                                                              \
        event_base_gettimeofday_cached(NULL, &cur_tv);                                  \
        event_user_debug((                                                              \
         "%s: evcb: %p , %s%s%s%s%s%s%s%s%s call %p, curtime(s/us):%d/%d",      \
         __func__,                                                                      \
         evcb,                                                                          \
         evcb->evcb_flags & EVLIST_TIMEOUT ? "EVLIST_TIMEOUT " : " ",                   \
         evcb->evcb_flags & EVLIST_INSERTED ? "EVLIST_INSERTED " : " ",                             \
         evcb->evcb_flags & EVLIST_SIGNAL ? "EVLIST_SIGNAL " : " ",                     \
         evcb->evcb_flags & EVLIST_ACTIVE ? "EVLIST_ACTIVE " : " ",                     \
         evcb->evcb_flags & EVLIST_INTERNAL ? "EVLIST_INTERNAL " : " ",                  \
         evcb->evcb_flags & EVLIST_ACTIVE_LATER ? "EVLIST_ACTIVE_LATER " : " ",          \
         evcb->evcb_flags & EVLIST_FINALIZING ? "EVLIST_FINALIZING " : " ",              \
         evcb->evcb_flags & EVLIST_INIT ? "EVLIST_INIT " : " ",                                 \
         tv ? "EV_TIMEOUT " : " ",                                                      \
         evcb,                                                               \
         cur_tv.tv_sec,                                                                 \
         cur_tv.tv_usec));                                                              \
        } while(0)


#define event_user_test_log_ev(ev, tv) do{                                                         \
    struct timeval cur_tv;                                                              \
        event_base_gettimeofday_cached(NULL, &cur_tv);                                  \
        event_user_debug((                                                              \
         "%s: event: %p (fd "EV_SOCK_FMT"), %s%s%s%scall %p, curtime(s/us):%d/%d",      \
         __func__,                                                                      \
         ev,                                                                            \
         EV_SOCK_ARG(ev->ev_fd),                                                        \
         ev->ev_events & EV_READ ? "EV_READ " : " ",                                    \
         ev->ev_events & EV_WRITE ? "EV_WRITE " : " ",                                  \
         ev->ev_events & EV_CLOSED ? "EV_CLOSED " : " ",                                \
         tv ? "EV_TIMEOUT " : " ",                                                      \
         ev->ev_callback,                                                               \
         cur_tv.tv_sec,                                                                 \
         cur_tv.tv_usec));                                                              \
	 event_user_test_log_evcb(event_to_event_callback(ev), tv);					\
        } while(0)
```
-------------------------

we call these log functions at the beginning of thses functions: 
```c
event_add
event_add_nolock_
event_del
event_del_
event_del_nolock_
event_queue_insert_inserted
event_queue_remove_inserted
event_queue_insert_active
event_queue_remove_active
```
-------------------------

--- logs collected
```
...
[debug] event_queue_remove_active: evcb: 0xaae5410 ,    EVLIST_ACTIVE    EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269625/736809
[debug] event_del: event: 0xaae5410 (fd 45),  EV_WRITE   call 0x7c4cf70, curtime(s/us):1487269625/737094
[debug] event_del: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269625/737170
[debug] event_del_: event: 0xaae5410 (fd 45),  EV_WRITE   call 0x7c4cf70, curtime(s/us):1487269625/737223
[debug] event_del_: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269625/737617
[debug] event_del_nolock_: event: 0xaae5410 (fd 45),  EV_WRITE   call 0x7c4cf70, curtime(s/us):1487269625/737698
[debug] event_del_nolock_: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269625/737760
...
[debug] event_del_nolock_: event: 0xaae5410 (fd 45),  EV_WRITE   call 0x7c4cf70, curtime(s/us):1487269625/874871
[debug] event_del_nolock_: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269625/874931
[debug] event_queue_remove_active: evcb: 0x6d332b0 ,  EVLIST_INSERTED  EVLIST_ACTIVE    EVLIST_INIT   call 0x6d332b0, curtime(s/us):1487269625/875019
...
[debug] event_queue_remove_inserted: evcb: 0xad024e0 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xad024e0, curtime(s/us):1487269627/632445
[debug] event_queue_remove_active: evcb: 0x6cfda80 ,    EVLIST_ACTIVE   EVLIST_FINALIZING EVLIST_INIT   call 0x6cfda80, curtime(s/us):1487269627/634394
[debug] event_queue_remove_active: evcb: 0xa8c6c10 ,    EVLIST_ACTIVE   EVLIST_FINALIZING EVLIST_INIT   call 0xa8c6c10, curtime(s/us):1487269627/635058
...
[debug] event_queue_insert_active: evcb: 0x6c54888 ,  EVLIST_INSERTED      EVLIST_INIT   call 0x6c54888, curtime(s/us):1487269627/658787
[debug] event_queue_insert_active: evcb: 0xab05410 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/658976
[debug] event_queue_insert_active: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269627/659050        -<<<===========0xaae5410  persistent and inserted again but with no EVLIST_INSERTED  flag set !!!
...
[debug] event_queue_insert_active: evcb: 0x6d21dd0 ,  EVLIST_INSERTED      EVLIST_INIT   call 0x6d21dd0, curtime(s/us):1487269627/684201
[debug] event_queue_remove_active: evcb: 0x6c54888 ,  EVLIST_INSERTED  EVLIST_ACTIVE    EVLIST_INIT   call 0x6c54888, curtime(s/us):1487269627/684312
[debug] event_add: event: 0x6cfd4b0 (fd -1), EV_READ    call 0x7e90790, curtime(s/us):1487269627/684589
[debug] event_add: evcb: 0x6cfd4b0 ,        EVLIST_INIT   call 0x6cfd4b0, curtime(s/us):1487269627/684664
[debug] event_add_nolock_: event: 0x6cfd4b0 (fd -1), EV_READ    call 0x7e90790, curtime(s/us):1487269627/684721
[debug] event_add_nolock_: evcb: 0x6cfd4b0 ,        EVLIST_INIT   call 0x6cfd4b0, curtime(s/us):1487269627/684782
[debug] event_queue_insert_inserted: event: 0x6cfd4b0 (fd -1), EV_READ    call 0x7e90790, curtime(s/us):1487269627/685126
[debug] event_queue_insert_inserted: evcb: 0x6cfd4b0 ,        EVLIST_INIT   call 0x6cfd4b0, curtime(s/us):1487269627/685201
[debug] event_add: event: 0x6ce5730 (fd 28), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/685302
[debug] event_add: evcb: 0x6ce5730 ,        EVLIST_INIT   call 0x6ce5730, curtime(s/us):1487269627/685373
[debug] event_add_nolock_: event: 0x6ce5730 (fd 28), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/685429
[debug] event_add_nolock_: evcb: 0x6ce5730 ,        EVLIST_INIT   call 0x6ce5730, curtime(s/us):1487269627/685482
[debug] event_queue_insert_inserted: event: 0x6ce5730 (fd 28), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/685566
[debug] event_queue_insert_inserted: evcb: 0x6ce5730 ,        EVLIST_INIT   call 0x6ce5730, curtime(s/us):1487269627/685619
[debug] event_add: event: 0x6dcb180 (fd -1), EV_READ    call 0x7e90790, curtime(s/us):1487269627/685818
[debug] event_add: evcb: 0x6dcb180 ,        EVLIST_INIT   call 0x6dcb180, curtime(s/us):1487269627/685883
[debug] event_add_nolock_: event: 0x6dcb180 (fd -1), EV_READ    call 0x7e90790, curtime(s/us):1487269627/685936
[debug] event_add_nolock_: evcb: 0x6dcb180 ,        EVLIST_INIT   call 0x6dcb180, curtime(s/us):1487269627/686036
[debug] event_queue_insert_inserted: event: 0x6dcb180 (fd -1), EV_READ    call 0x7e90790, curtime(s/us):1487269627/686091
[debug] event_queue_insert_inserted: evcb: 0x6dcb180 ,        EVLIST_INIT   call 0x6dcb180, curtime(s/us):1487269627/686143
...
[debug] event_queue_remove_active: evcb: 0xab05410 ,  EVLIST_INSERTED  EVLIST_ACTIVE    EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/738224
[debug] event_del: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/738444
[debug] event_del: evcb: 0xab05410 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/738512
[debug] event_del_: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/738565
[debug] event_del_: evcb: 0xab05410 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/738619
[debug] event_del_nolock_: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/738676
[debug] event_del_nolock_: evcb: 0xab05410 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/738732
[debug] event_queue_remove_inserted: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/738784
[debug] event_queue_remove_inserted: evcb: 0xab05410 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/738866
[debug] event_add: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/738985
[debug] event_add: evcb: 0xab05410 ,        EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/739038
[debug] event_add_nolock_: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/739089
[debug] event_add_nolock_: evcb: 0xab05410 ,        EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/739138
[debug] event_queue_insert_inserted: event: 0xab05410 (fd 64), EV_READ    call 0x7c4da90, curtime(s/us):1487269627/739200
[debug] event_queue_insert_inserted: evcb: 0xab05410 ,        EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/739252

[debug] event_queue_remove_active: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269627/739309
[err] event_queue_remove_active: 0xaae5410 not on queue 8    -<<<===========0xaae5410 
```

It fails when `0xaae5410` is removed from the active queue while its `EVLIST_ACTIVE` is not set.
Seems some negative data race exists and the `EVLIST_ACTIVE` mask or `EVLIST_INSERTED` mask is set in a wrong order , thus vialoating the implied happen-before relation?

We compared with other callbacks from log seq and found the general pattern:
`EVLIST_INSERTED`  `EVLIST_ACTIVE`   `EVLIST_INIT`  three masks were set when remove like: 
```
[debug] event_queue_insert_active: evcb: 0xab05410 ,  EVLIST_INSERTED      EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/658976
[debug] event_queue_remove_active: evcb: 0xab05410 ,  EVLIST_INSERTED  EVLIST_ACTIVE    EVLIST_INIT   call 0xab05410, curtime(s/us):1487269627/738224 
```

or  `EVLIST_ACTIVE`  `EVLIST_FINALIZING` `EVLIST_INIT` three masks were set
```
[debug] event_queue_insert_active: evcb: 0x6ec9ef0 ,        EVLIST_INIT   call 0x6ec9ef0, curtime(s/us):1487269627/473269
[debug] event_queue_remove_active: evcb: 0x6ec9ef0 ,    EVLIST_ACTIVE   EVLIST_FINALIZING EVLIST_INIT   call 0x6ec9ef0, curtime(s/us):1487269627/654246
```

but the failed one:
```
[debug] event_queue_insert_active: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269627/659050
[debug] event_queue_remove_active: evcb: 0xaae5410 ,        EVLIST_INIT   call 0xaae5410, curtime(s/us):1487269627/739309
[err] event_queue_remove_active: 0xaae5410 not on queue 8
```

**os version:**
```
[root@localhost ~]# cat /etc/issue
CentOS release 6.5 (Final)
Kernel \r on an \m

[root@localhost ~]# uname -a
Linux localhost.localdomain 2.6.32-431.el6.x86_64 #1 SMP Fri Nov 22 03:15:09 UTC 2013 x86_64 x86_64 x86_64 GNU/Linux
```

Any suggestions and ideas will be appreciated, please help.

**Comment by azat** [comment]
Can you please try with debug checks enabled (I hope that you compile with them):
```
    # put this at the top of main()
    event_enable_debug_mode();
```

P.S. preferable version for testing is **2.1.8**

**Comment by Zourva** [comment]
Thank you, @azat . I'll try your suggestion using 2.1.8 and thanks for your edit.

**Comment by Zourva** [comment]
We reproduced the result with `event_enable_debug_mode() `on  2.1.8.  The collected log size is about 5.1GB. Here's the logs related to the failed event (`fd 216` according to the call trace): 

```

[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 216),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 216), res 4, callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 216), callback 0x2b13695b8f70
...
[debug] event_add: event: 0x2b13881d8aa0 (fd 43),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 43), res 4, callback 0x2b13695b8f70
[debug] event_process_active: event: 0x2b13881d8aa0,  EV_WRITE  call 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 43), callback 0x2b13695b8f70
[debug] event_add: event: 0x2b13881d8aa0 (fd 43),  EV_WRITE   call 0x2b13695b8f70
[debug] event_active: 0x2b13881d8aa0 (fd 43), res 4, callback 0x2b13695b8f70
[debug] event_process_active: event: 0x2b13881d8aa0,  EV_WRITE  call 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 43), callback 0x2b13695b8f70
[debug] event_del: 0x2b13881d8aa0 (fd 43), callback 0x2b13695b8f70
[err] event_assign called on an already added event 0x2b13881d8aa0 (events: 0x0, fd: 0, flags: 0x0)

```
It seems that `event_process_active` is not called on this event when fd is 216. 
Will this error lead to the early 'not on queue 8 ' error ? 　

call trace:
```
#5  0x00002b13695c80a1 in event_exit (errcode=-559030611) at log.c:102
#6  0x00002b13695c816c in event_errx (eval=-559030611, fmt=<value optimized out>) at log.c:162
#7  0x00002b13695bded3 in event_assign (ev=0x2b13881d8aa0, base=0x2b137c0014d0, fd=216, events=84, callback=0x2b13695b8f70 <bufferevent_writecb>, arg=0x2b13881d8a10)
    at event.c:2072
#8  0x00002b13695b9655 in bufferevent_socket_new (base=0x2b137c0014d0, fd=216, options=1) at bufferevent_sock.c:367
#9  0x00002b1369a0378e in _evhtp_connection_accept (evbase=0x2b137c0014d0, connection=0x2b13885e9470) at /root/workbench/rpm_pkgs/libevhtp-1.2.10/evhtp.c:2017
#10 0x00002b1369a04f41 in _evhtp_accept_cb (serv=<value optimized out>, fd=216, s=0x2b1380601b70, sl=16, arg=0x2b137c001e20)
    at /root/workbench/rpm_pkgs/libevhtp-1.2.10/evhtp.c:2186
#11 0x00002b13695c7b9a in listener_read_cb (fd=21, what=<value optimized out>, p=0x2b137c002140) at listener.c:418
#12 0x00002b13695c01a7 in event_persist_closure (base=0x2b137c0014d0, activeq=0x2b137c001b60, max_to_process=2147483647, endtime=0x0) at event.c:1580
#13 event_process_active_single_queue (base=0x2b137c0014d0, activeq=0x2b137c001b60, max_to_process=2147483647, endtime=0x0) at event.c:1639
#14 0x00002b13695c09af in event_process_active (base=0x2b137c0014d0, flags=<value optimized out>) at event.c:1738
#15 event_base_loop (base=0x2b137c0014d0, flags=<value optimized out>) at event.c:1961
```

_evhtp_xxx are functions from @ellzey libevhtp

**Comment by azat** [comment]
FWIW here is some information about this error: nmathewson/Libevent#141

**Comment by azat** [comment]
And indeed this error (calling event_assing() on already active/pending event) can lead to this error, and you shouldn't have code that do this.

Look, in `event_queue_remove_active()` we are checking `evcb_flags/ev_flags` for `EVLIST_ACTIVE` but `event_assign()` reset `ev_flags` to `EVLIST_INIT`.

**Comment by Zourva** [comment]
Hi azat, thanks for your reply. I check the link and the code again, finding that event_assign() is called against a newly-malloc'ed event, which is why the log shows `event_assign called on an already added event 0x2b13881d8aa0 **(events: 0x0, fd: 0, flags: 0x0)**`. No flags are set because it's not initialized yet. Since it's not initialized(thus no reset happens), why is it confilict ?

```
struct bufferevent *
bufferevent_socket_new(struct event_base *base, evutil_socket_t fd,
    int options)
{
	struct bufferevent_private *bufev_p;
	struct bufferevent *bufev;

#ifdef _WIN32
	if (base && event_base_get_iocp_(base))
		return bufferevent_async_new_(base, fd, options);
#endif

	if ((bufev_p = mm_calloc(1, sizeof(struct bufferevent_private)))== NULL)
		return NULL;

	if (bufferevent_init_common_(bufev_p, base, &bufferevent_ops_socket,
				    options) < 0) {
		mm_free(bufev_p);                    ---- this function doesn't update ev_read/ev_write either
		return NULL;
	}
	bufev = &bufev_p->bev;                    ---- bufev_p is a fresh object, just malloc'ed above
	evbuffer_set_flags(bufev->output, EVBUFFER_FLAG_DRAINS_TO_FD);

	event_assign(&bufev->ev_read, bufev->ev_base, fd,
	    EV_READ|EV_PERSIST|EV_FINALIZE, bufferevent_readcb, bufev);
	event_assign(&bufev->ev_write, bufev->ev_base, fd,                   ---- the check fails inside this call
	    EV_WRITE|EV_PERSIST|EV_FINALIZE, bufferevent_writecb, bufev);

	evbuffer_add_cb(bufev->output, bufferevent_socket_outbuf_cb, bufev);

	evbuffer_freeze(bufev->input, 0);
	evbuffer_freeze(bufev->output, 1);

	return bufev;
}
```

**Comment by azat** [comment]
Hm, didn't looked at backtrace, and indeed this is a new event, that shouldn't have any problems.

Looks like a corruption to me somewhere else, i.e. if this region of memory freed and malloc thinks so, for example if you call `free()` instead of `event_free()`. Can you try compile with `-fsanitize=address` or running under `valgrind ...`?

`-fsanitize=address` has less cpu overhead, so it is preferable.

**Comment by Zourva** [comment]
Hi azat, we recompiled with `-fsanitize=address` and `-fsanitize=leak` and run. Output showed `heap-use-after-free` as you said. 

We run using valgrind and helgrind but it never showed. 
We have a struct holding the request object pointer and get queued:
```
typedef struct {
    ...
    evhtp_request* req;
}session;
```
We checked our usage of `evhtp` and found that  we missed a lock in `session`!!! 
Both valgrind and helgrind failed to show the mistake, we failed to. It's such a common case of race access though ! T.T 

Thank you very much for your help.
